app/src/main/java/com/fstyle/structure_android/utils/Validator.java

```diff
@@ -0,0 +1,277 @@
+                    e.printStackTrace();
+                }
+                if (mErrorTraces.indexOfKey(type) != -1) {
+                    mMessageInvalids.add(rule.message());
+                }
+            }
+        }
+    }
+
+    public <T extends BaseModel> Validator prepare(T model) {
+        mModelCache = model.clone();
+        return this;
+    }
+
+    public <T extends BaseModel> void validate(T model, boolean onlyValidateChange) {
+        if (model == null || mModelCache == null) {
+            return;
+        }
+
+        for (Field field : model.getClass().getDeclaredFields()) {
+            Validation annotation = field.getAnnotation(Validation.class);
+            if (annotation == null) {
+                continue;
+            }
+            Rule[] rules = annotation.value();
+            Optional optional = field.getAnnotation(Optional.class);
+            boolean isOptional = optional != null;
+            field.setAccessible(true);
+
+            try {
+                Object cache = field.get(mModelCache);
+                Object real = field.get(model);
+                // detect when data has changed only if onlyValidateTheChange is true.
+                // Otherwise, always run validation on all of fields .
+                if (!onlyValidateChange || (cache == null && real != null) || (cache != null
+                        && cache.equals(real))) {
+                    validate(real, rules, isOptional);
+                }
+            } catch (IllegalAccessException e) {
+                Log.e(TAG, "validate: ", e);
+            }
+            mModelCache = model.clone();
+        }
+    }
+
+    @StringRes
+    public int getErrorByTypeInt(@ValidType int type) {
+        int index = mErrorTraces.indexOfValue(type);
+        if (index == -1) {
+            return R.string.empty;
+        }
+        return mErrorTraces.keyAt(index);
+    }
+
+    public String getErrorByType(@ValidType int type) {
+        int index = mErrorTraces.indexOfValue(type);
+        if (index == -1 || mErrorTraces.keyAt(index) == -1) {
+            return null;
+        }
+        return mContext.getString(mErrorTraces.keyAt(index));
+    }
+
+    public void initNGWordPattern() {
+        Observable.create(new Observable.OnSubscribe<Pattern>() {
+            @Override
+            public void call(Subscriber<? super Pattern> subscriber) {
+                try {
+                    BufferedReader reader = new BufferedReader(
+                            new InputStreamReader(mContext.getAssets().open("ng-word")));
+                    StringBuilder buffer = new StringBuilder();
+                    String line;
+                    while ((line = reader.readLine()) != null) {
+                        if (buffer.length() == 0) {
+                            buffer.append("(").append(line.toLowerCase());
+                        } else {
+                            buffer.append("|").append(line.toLowerCase());
```

Comment:
WARNING: [AppendCharacterWithChar] Avoid appending characters as strings in StringBuffer.append.
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/strings.html#AppendCharacterWithChar